Protocols: Migrate WebDAV XML parsing to Python's ElementTree

akselaase opened the pull request:
As described in #5700, the parser for the WebDAV server responses had issues with namespace handling. This caused Rucio to misinterpret responses from for instance the Apache DAV module. Changing the implementation to use Python's `xml.etree.ElementTree` fixes this issue.

This PR also includes a fix for #5308, in which Rucio doesn't properly handle DAV servers with path prefixes.

<!-- Please read https://rucio.cern.ch/documentation/contributing before submitting a pull request -->

akselaase commented on setuputil.py:
```diff
@@ -18,6 +18,12 @@
+
+if sys.version_info[:2] == (3, 6):
+    _extra_client_requirements.append('dataclasses')
+    _extra_server_requirements.append('dataclasses')
```
TODO: rename client->common and don't append to server list.

joeldierkes replied:
Info: We use markers in the requirements file to include packages in specific environments: https://github.com/rucio/rucio/commit/c301af509ba7eed5e34b644a68db39fcc2fa7553 : e.g. the mock package that got removed.

This work because we parse the requirements file with the `pkg_resources` package. [https://setuptools.pypa.io/en/latest/pkg_resources.html](https://setuptools.pypa.io/en/latest/pkg_resources.html#requirements-parsing)

As far as I can tell, you don't need to introduce the different lists. All elements in the install_requires of the client are also in the server.
```python
server_requirements_table = {
    'install_requires': clients_requirements_table['install_requires'] + [
```

joeldierkes commented on lib/rucio/rse/protocols/webdav.py:
```diff
@@ -89,56 +90,63 @@ def __len__(self):
+        else:
+            size = int(xml_size.text)
+
+        return _PropfindFile(href=href, size=size)
```
How about using a `classmethod`?

joeldierkes commented on lib/rucio/rse/protocols/webdav.py:
```diff
@@ -89,56 +90,63 @@ def __len__(self):
+
+    files: Tuple[_PropfindFile]
+
+    @staticmethod
```
+1

joeldierkes commented on lib/rucio/rse/protocols/webdav.py:
```diff
@@ -483,12 +496,28 @@ def stat(self, path):
+
+            else:
+                raise exception.ServiceUnavailable("WebDAV response didn't include content length for requested path.")
+
```
As far as I see it this does not has to be inside the outer try?

joeldierkes commented on lib/rucio/rse/protocols/webdav.py:
```diff
@@ -89,56 +90,63 @@ def __len__(self):
 
-    """ Parser to parse XML output for PROPFIND ."""
+    href: Optional[str]
+    size: Optional[int]
```
Is there a use case where we recieve one of `href`, `size`, but not the other?

From my perspective these are tighly coulped, I'd rather see a `None` there opposed to `_PropfindFile(None, 22)` or `_PropfindFile(path, None)`, which both does not make sense?

At least `href` should not be Optional...

akselaase replied:
It's just because I can't guarantee what the WebDAV server responds with. I agree that `href` could be made mandatory (it is in the RFC at least), so I'll change it to throw an exception if it is missing from the response.

joeldierkes commented on requirements.txt:
```diff
@@ -4,6 +4,7 @@ urllib3>=1.24.2,<=1.26.8                                    # HTTP library with
 dogpile.cache>=1.1.2,<=1.1.5                                # Caching API plugins (1.1.2 is the first version to support pymemcache)
 tabulate~=0.8.0                                             # Pretty-print tabular data
 jsonschema~=3.2.0                                           # For JSON schema validation (Policy modules)
+dataclasses==0.8.0; python_version == '3.6'                  # For dataclass support in 3.6
```
Comment alignment, plus the comment itself is useless. I am not in favor of documenting it because everyhting else is when it adds not value.

joeldierkes reviewed (changes requested):
It's an _interesting_ PR...

Small nits.

I'd also propose to split the commits into two:
- Introducing dataclasses
- Things related to the webdav stuff

That way we can revert the webdav stuff in case of an error, without removing the precious dataclasses.

akselaase committed:
Core & Internals: add dependency on "dataclasses" for python 3.6; Related to #5700

akselaase committed:
Protocols: Migrate WebDAV XML parsing to Python's ElementTree; Fix #5700, fix #5308

As described in #5700, the parser for the WebDAV server responses had
issues with namespace handling. This caused Rucio to misinterpret
responses from for instance the Apache DAV module. Changing the
implementation to use Python's xml.etree.ElementTree fixes this issue.

This PR also includes a fix for #5308, in which Rucio doesn't properly
handle DAV servers with path prefixes.

bari12 reviewed (approved):
👍 Looks good. I will only include it in 1.30 though, due to the added dependency.
